### ISSUE TITLE
Meta-information dropout

### ISSUE THREAD

**Issue by RaymondLi0** [bigcode-project/Megatron-LM#30]
Add document meta-information at the beginning (filename, filepath, repo name, number of stars...)
Also there should be a dropout for each of these fields.

- [ ] Add meta-information during preprocessing at the beginning of each document
- [ ] Dropout this information when loading examples during training. Use a seed.

Optional: at the beginning of the preprocessed document, pre-pend the number of tokens that contain the meta information. That would allow to detokenize only the meta-information instead of the whole document during training.

On issue would be that preprocessed samples of length 2048 would get shorter if the meta-information is removed. Possible workaround:
- add padding. Not ideal because it would add computational waste.
- build index-mappings with longer sequences (e.g. 2148 instead of 2048) and truncate as needed: waste less compute, but waste some data. Would still need padding in some cases

Cleaner solution: tokenize document content and meta-information strings separately. Let N be the number of meta-information field that can be dropped-out. We create N+1 preprocessed flies. When loading data at training time, read through these N+1 files in parallel and add the meta-info with dropout. More implementation work needed in this solution

**Comment by RaymondLi0** [comment]
Maybe start with the simpler implementation and see how much padding we get. A very low amount (<1%) is acceptable.

**Comment by RaymondLi0** [comment]
Since we won't do a lot of pass over the same data, we could consider doing dropout during preprocessing (instead of on-the-fly during training). This would not require any change in megatron
